### ISSUE TITLE
[Question] Can this be used with Frigate+ models?

### ISSUE THREAD

**Issue by BlackBeltPanda** [dbro/frigate-detector-edgetpu-yolo9#1]
I just subscribed to Frigate+ and saw that my edgeTPU isn't compatible with the yolov9 model, so before I request a mobiledet model I figured I'd see if there's a way to use the edgeTPU and came across this repo. Is it possible to use this plugin with the yolov9 model from Frigate+?

**Comment by bexem** [comment]
That would be great, hopefully @dbro (thank you!) might have the answer. 
I've just finished installing it, on unraid and with USB Coral, so far so good with low resource usage although it's evening here so I will only be able to fully test during the coming days.

**Comment by dbro** [comment]
Hello
I am not a Frigate+ subscriber, so I don't have access to those model files. It might be possible to convert the Frigate+ yolo v9 ONNX model to TFLite and then to edgetpu. If you want to try to follow the example in this Google Colab notebook and do that, let me know how it goes.

https://colab.research.google.com/drive/1n3sCrsVWJKu2H5KjgUyl3akIUYykSYaP?usp=sharing

Note that the accuracy of the model produced by this version is OK, but could be better. I am working on a new version of the model that will hopefully have fewer false negatives. See here https://www.reddit.com/r/frigate_nvr/comments/1ohgzsv/reducing_false_positives_by_running_yolo_v9_on/ in particular the comment from yesterday that starts with "Some good news"

(sorry for the delayed response, I didn't get an alert until today's message from @bexem)

**Comment by AnderssonPeter** [comment]
It would be nice if the readme included instruction on how to convert a model in a optimal way to be used with this solution

**Comment by dbro** [comment]
Hi @AnderssonPeter here are two colab notebooks that show how to create and convert the models. They are still a bit messy, because I expect most people will just download the ready-made model file for Google Coral instead of going through the trouble to fine-tune their own.

1) https://colab.research.google.com/drive/1RAZqCHuDWFdFxqFpL78OVfZQvnGoM6rH?usp=sharing

2) https://colab.research.google.com/drive/1AdFfPn1zYj86e_n2Bq3hvvNbIlyn8Vei?usp=sharing

The first notebook starts with the pre-trained YOLO model and changes the activation function to ReLU and reduces the number of COCO classes to the 17 that are relevant for Frigate. Note that there is a subsampling parameter of 10% which allows this to run quickly enough to complete on the Google Colab free tier. If you want to use the full COCO dataset, then adjust that parameter to 100 and expect it to take about 1 hour per epoch. I ran it for 10 epochs with the full data set to create the model file that can be downloaded here. It looks like after 10 epochs, the model's validation scores get a bit worse, perhaps an indication of overfitting.

The second notebook takes the fine-tuned model from the first notebook and converts it to a version of TFLite that can run on Google Coral. It has some intermediate checks which were helpful when investigating what was causing quantization problems. As you can see, there are still 3 operations (out of 339) that do not run on the TPU, it would be nice to get that number to zero but I wasn't able to figure out where in the boxes pipeline the issue was. It might run a bit faster if that were taken care of.

**Comment by dbro** [comment]
Thanks @Alan01252 you are right, and the current model does run entirely on the TPU. After changing the order of the dimensions of the boxes tensor to be BNC (1, 2100, 64) instead of BCN (1, 64, 2100) it was able to do everything on the TPU. The other smaller tensor with classes information (1, 17, 2100) never had this issue, so I think your assumption about them getting too big fits with the observed behavior.

The output from the edgetpu-compiler now looks like this:

Running: edgetpu_compiler yolov9-s-relu6-best_320_int8.tflite -o . --show_operations --search_delegate --delegate_search_step 30 --timeout_sec 180
Edge TPU Compiler version 16.0.384591198
Searching for valid delegate with step 30
Try to compile segment with 333 ops
Started a compilation timeout timer of 180 seconds.

Model compiled successfully in 6822 ms.

Input model: yolov9-s-relu6-best_320_int8.tflite
Input size: 7.23MiB
Output model: ./yolov9-s-relu6-best_320_int8_edgetpu.tflite
Output size: 7.80MiB
On-chip memory used for caching model parameters: 7.23MiB
On-chip memory remaining for caching model parameters: 768.00B
Off-chip memory used for streaming uncached model parameters: 385.38KiB
Number of Edge TPU subgraphs: 1
Total number of operations: 333
Operation log: ./yolov9-s-relu6-best_320_int8_edgetpu.log

Operator                       Count      Status

RESHAPE                        6          Mapped to Edge TPU
RESIZE_NEAREST_NEIGHBOR        2          Mapped to Edge TPU
CONV_2D                        203        Mapped to Edge TPU
SPLIT                          6          Mapped to Edge TPU
ADD                            48         Mapped to Edge TPU
PAD                            2          Mapped to Edge TPU
STRIDED_SLICE                  16         Mapped to Edge TPU
AVERAGE_POOL_2D                5          Mapped to Edge TPU
MAXIMUM                        3          Mapped to Edge TPU
QUANTIZE                       1          Mapped to Edge TPU
MAX_POOL_2D                    3          Mapped to Edge TPU
MINIMUM                        3          Mapped to Edge TPU
CONCATENATION                  35         Mapped to Edge TPU
Compilation child process completed within timeout period.
Compilation succeeded!

**Comment by Alan01252** [comment]
This pleases me greatly :) Excellent work!
